test/Generics/protocol_where_clause.swift

```diff
@@ -65,3 +65,12 @@ struct BadConcreteNestedConforms: NestedConforms {
   // expected-error@-1 {{type 'ConcreteParentNonFoo2.T' (aka 'Float') does not conform to protocol 'Foo2'}}
   typealias U = ConcreteParentNonFoo2
 }
+
+// <rdar://problem/31041997> Some where-clause requirements aren't checked on conforming types
+protocol P1 {}
+protocol P2 { associatedtype T }
+protocol P3 { associatedtype U }
+protocol P4: P3 where U: P2, U.T: P1 {}
```

Comment:
Oh, you're right. Well then! I'll see if I can rip out the hack in #8428 now; it might require better handling of concrete types in `GenericSignature`.